### PR TITLE
mgr/stats: be resilient to offline rank0 MDS

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/50033

### CODE REVIEW

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 1]
```diff
@@ -159,6 +159,14 @@ def notify(self, cmdtag):
             self.log.debug("client_metadata={0}, to_purge={1}".format(
                 self.client_metadata['metadata'], self.client_metadata['to_purge']))
 
+    def re_register_queries(self):
+        #re-register user queries
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:163]: It might not be a good idea to (re)register queries in nofity() since this is invoked by ceph-mgr thread (probably finisher thread) which might keep it blocked for long (think, registering ~100 queries to the MDS).

Comment (vshankar) [reply src/pybind/mgr/stats/fs/perf_stats.py:163]: Just  FYI, you can schedule a timer on notify() when rank0 changes to re-register queries.

Comment (joscollin) [reply src/pybind/mgr/stats/fs/perf_stats.py:163]: Fixed the review comments. Please check.

File: src/pybind/mgr/stats/module.py [hunk 2]
```diff
@@ -25,14 +26,35 @@ class Module(MgrModule):
...
         raise NotImplementedError(cmd['prefix'])
+
+    def get_rank0_mds_name(self, fsmap):
+        for fs in fsmap['filesystems']:
+            mds_map = fs['mdsmap']
+            if mds_map is not None:
+                for mds_id, mds_status in mds_map['info'].items():
+                    if mds_status['rank'] == 0:
+                        return mds_status['name']
+        raise RuntimeError("Failed to find a rank0 mds in fsmap")
```
Comment (vshankar) [inline src/pybind/mgr/stats/module.py:60]: maybe log and ignore? we'll get notified with rank0 soon anyway.

File: src/pybind/mgr/stats/module.py [hunk 3]
```diff
@@ -25,14 +26,35 @@ class Module(MgrModule):
...
+                log.warn(e)
 
     def handle_command(self, inbuf, cmd):
         prefix = cmd['prefix']
         # only supported command is `fs perf stats` right now
         if prefix.startswith('fs perf stats'):
             return self.fs_perf_stats.get_perf_data(cmd)
         raise NotImplementedError(cmd['prefix'])
+
+    def get_rank0_mds_name(self, fsmap):
```
Comment (vshankar) [inline src/pybind/mgr/stats/module.py:53]: can be a static method.

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 4]
```diff
@@ -159,6 +159,14 @@ def notify(self, cmdtag):
...
+        for filter_spec in list(self.user_queries.keys()):
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:164]: nit: add a debug log before re-registering.

**Review by vshankar — CHANGES_REQUESTED** [review]

**Comment by vshankar** [comment]
@joscollin I'll take a look the the fixes. This would need a test case in test_mds_metrics.py

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 5]
```diff
@@ -161,6 +161,7 @@ def notify(self, cmdtag):
 
     def re_register_queries(self):
         #re-register user queries
+        self.log.debug("reregistering queries...")
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:164]: nit: also add debug logs for the queries itself.

File: src/pybind/mgr/stats/module.py [hunk 6]
```diff
@@ -26,22 +29,16 @@ class Module(MgrModule):
     def __init__(self, *args, **kwargs):
         super(Module, self).__init__(*args, **kwargs)
         self.fs_perf_stats = FSPerfStats(self)
-        try:
-            self.prev_mds_name_rank0 = self.get_rank0_mds_name(self.get('fs_map'))
-        except Exception:
-            raise
+        self.prev_rank0_name = Module.get_rank0_mds_name(self.get('fs_map'))
```
Comment (vshankar) [inline src/pybind/mgr/stats/module.py:32]: I think I already commented on this -- rank 0 should be fetched in notify()

File: src/pybind/mgr/stats/module.py [hunk 7]
```diff
@@ -25,14 +26,35 @@ class Module(MgrModule):
     def __init__(self, *args, **kwargs):
         super(Module, self).__init__(*args, **kwargs)
         self.fs_perf_stats = FSPerfStats(self)
+        try:
+            self.prev_mds_name_rank0 = self.get_rank0_mds_name(self.get('fs_map'))
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/mgr/stats/module.py]: were you able to check this? what happens when rank0 is unavailable when adding a user query.

**Review by vshankar — CHANGES_REQUESTED** [review]
@joscollin I would prefer if the most of the stuff is done in per_stats.py rather than in module.py.

File: src/pybind/mgr/stats/module.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (joscollin) [reply src/pybind/mgr/stats/module.py]: If rank0 becomes unavailable at some point, `get_rank0_mds_name` returns `None` and skips the query reregistering. Will push the latest code today for you to check.

File: src/pybind/mgr/stats/module.py [hunk 8]
```diff
@@ -25,10 +26,16 @@ class Module(MgrModule):
     def __init__(self, *args, **kwargs):
         super(Module, self).__init__(*args, **kwargs)
         self.fs_perf_stats = FSPerfStats(self)
+        self.prev_rank0_name = None
 
     def notify(self, notify_type, notify_id):
         if notify_type == "command":
             self.fs_perf_stats.notify(notify_id)
+        elif notify_type == "fs_map":
```
Comment (vshankar) [inline src/pybind/mgr/stats/module.py:34]: why? can't this be done in FSPerfStats::notify?

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 9]
```diff
@@ -159,6 +159,26 @@ def notify(self, cmdtag):
...
 
+    def re_register_queries(self):
+        self.log.debug("reregistering queries...")
+        for filter_spec in list(self.user_queries.keys()):
+            with self.lock:
+                user_query = self.user_queries[filter_spec]
+                user_query[QUERY_IDS] = self.register_mds_perf_query(filter_spec)
+                user_query[GLOBAL_QUERY_ID] = self.register_global_perf_query(filter_spec)
+                user_query[QUERY_LAST_REQUEST] = datetime.now()
+                self.log.debug("reregistered query: {}".format(user_query))
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:170]: nit: log before registering

**Review by vshankar — CHANGES_REQUESTED** [review]
@joscollin please mark the comments that have been addressed as "resolved". It becomes hard to follow otherwise.

File: src/pybind/mgr/stats/module.py (hunk 8 above)
Comment (joscollin) [reply src/pybind/mgr/stats/module.py:34]: Because FSPerfStats::notify doesn't seem to be a callback. Having the `notify_id` empty for `fs_map` notify_type, I could still move this block of code to a function say FSPerfStats::handle_fsmap_change(). Any better suggestions?

Comment (joscollin) [reply src/pybind/mgr/stats/module.py:34]: Another issue is `self.get('fs_map')` should be called via `module` then.

Comment (vshankar) [reply src/pybind/mgr/stats/module.py:34]: > Because FSPerfStats::notify doesn't seem to be a callback. Having the `notify_id` empty for `fs_map` notify_type, I could still move this block of code to a function say FSPerfStats::handle_fsmap_change(). Any better suggestions?

You could have FSPerfStats::notify_cmd() and FSPerfStats::notify_fsmap().

**Comment by vshankar** [comment]
@joscollin also add a test case for this.

**Comment by vshankar** [comment]
ceph-mgr disallows re-registering the same query, i.e., the query call never reaches the new rank zero mds. See: src/mgr/MetricCollector.cc::add_query(). The `notify` flag is not set if the query is already registered.

One way to fix this would be to force the query registry by passing something like a `reregister` flag when registering query. ceph-mgr can then use this as an override. Maybe it should preserve and return the existing `query-id`.

@joscollin What do you think? Also, just a small note -- please involve the core team for reviews for any changes to ceph-mgr (MetricCollector.cc, etc..).

**Comment by joscollin** [comment]
> ceph-mgr disallows re-registering the same query, i.e., the query call never reaches the new rank zero mds. See: src/mgr/MetricCollector.cc::add_query(). The `notify` flag is not set if the query is already registered.
> 
> One way to fix this would be to force the query registry by passing something like a `reregister` flag when registering query. ceph-mgr can then use this as an override. Maybe it should preserve and return the existing `query-id`.
> 
> @joscollin What do you think? 

Yes, I agree the same. As we discussed earlier, the reregistering should go through the `handle_query_updated()` instead of skipping it (notify=0). We could pass a flag from the perf_stats->re_register_queries() and set the `notify`, so that the call would reach `_send_configure()` call for the mds.

> Also, just a small note -- please involve the core team for reviews for any changes to ceph-mgr (MetricCollector.cc, etc..).
Will do. Thanks.

**Comment by joscollin** [comment]
> ceph-mgr disallows re-registering the same query, i.e., the query call never reaches the new rank zero mds. See: src/mgr/MetricCollector.cc::add_query(). The `notify` flag is not set if the query is already registered.
> 
> One way to fix this would be to force the query registry by passing something like a `reregister` flag when registering query. ceph-mgr can then use this as an override. Maybe it should preserve and return the existing `query-id`.
> 
> @joscollin What do you think? Also, just a small note -- please involve the core team for reviews for any changes to ceph-mgr (MetricCollector.cc, etc..).

@vshankar Updated the PR based on the comments. Please check.
Now it hits `_send_configure()` for the new rank0 mds. But it doesn't resolve the stale metrics issue (the waiting continues).

**Comment by vshankar** [comment]
> @vshankar Updated the PR based on the comments. Please check.
> Now it hits `_send_configure()` for the new rank0 mds. But it doesn't resolve the stale metrics issue (the wait for getting the latest metrics continues).

Nice. That fixes one part of the problem (I still need to review the changes though). Do we know why metrics are not forwarded by new rank 0 mds to ceph-mgr? (does mds logs tell anything)

**Comment by joscollin** [comment]
> > @vshankar Updated the PR based on the comments. Please check.
> > Now it hits `_send_configure()` for the new rank0 mds. But it doesn't resolve the stale metrics issue (the wait for getting the latest metrics continues).
> 
> Nice. That fixes one part of the problem (I still need to review the changes though). Do we know why metrics are not forwarded by new rank 0 mds to ceph-mgr? (does mds logs tell anything)

The metrics are being forwarded by `DaemonServer::_send_configure()` almost immediately. But the [MgrClient::handle_mgr_configure()](https://github.com/ceph/ceph/blob/master/src/mgr/MgrClient.cc#L446) is being called after long waiting. I didn't find the actual cause yet, but I could see some lookups going on during the wait period `2021-09-29T18:28:52.920+0530 7f761957e640 20 mds.0.cache.dir(0x100) lookup (head, 'stray1')`. Will check further.

**Comment by vshankar** [comment]
> The metrics are being forwarded by `DaemonServer::_send_configure()` almost immediately. But the [MgrClient::handle_mgr_configure()](https://github.com/ceph/ceph/blob/master/src/mgr/MgrClient.cc#L446) is being called after long waiting. I didn't find the actual cause yet, but I could see some lookups going on during the wait period `2021-09-29T18:28:52.920+0530 7f761957e640 20 mds.0.cache.dir(0x100) lookup (head, 'stray1')`. Will check further.

Could you share the logs where you saw handle_mgr_configure called with a huge delay? _send_configure() sends a MMgrConfigure message to the client and that should call handle_mgr_configure without much delay unless the message is stuck in a queue or something.

The lookup you are seeing should really not be causing any sort of delays.

**Comment by joscollin** [comment]
> > The metrics are being forwarded by `DaemonServer::_send_configure()` almost immediately. But the [MgrClient::handle_mgr_configure()](https://github.com/ceph/ceph/blob/master/src/mgr/MgrClient.cc#L446) is being called after long waiting. I didn't find the actual cause yet, but I could see some lookups going on during the wait period `2021-09-29T18:28:52.920+0530 7f761957e640 20 mds.0.cache.dir(0x100) lookup (head, 'stray1')`. Will check further.
> 
> Could you share the logs where you saw handle_mgr_configure called with a huge delay? _send_configure() sends a MMgrConfigure message to the client and that should call handle_mgr_configure without much delay unless the message is stuck in a queue or something.
> 
> The lookup you are seeing should really not be causing any sort of delays.

The log files are large size. So I've shared it in the drive. 

Failover of rank0 from mds.c to mds.b. The failover time in the logs is `Thursday 30 September 2021 02:44:18 PM IST`. 

Debug keywords in mgr.x.log are:
1. `query found in the queries map`
2. `_send_configure hit`

Search for them^.

Then search for `handle_mgr_configure` in mds.b.log after `rank0 is mds.b`. This is at `2021-09-30T15:11:29` which is the huge delay from the failover time.

File: src/mgr/ActivePyModules.cc [hunk 10]
```diff
@@ -1411,6 +1412,11 @@ PyObject *ActivePyModules::get_mds_perf_counters(MetricQueryID query_id)
     f.close_section(); // i
   }
   f.close_section(); // counters
+
+  f.open_array_section("timestamp");
```
Comment (vshankar) [inline src/mgr/ActivePyModules.cc:1416]: nit: last_updated?

File: src/mgr/ActivePyModules.cc [hunk 11]
```diff
@@ -1355,9 +1355,10 @@ PyObject *ActivePyModules::get_osd_perf_counters(MetricQueryID query_id)
 
 MetricQueryID ActivePyModules::add_mds_perf_query(
     const MDSPerfMetricQuery &query,
-    const std::optional<MDSPerfMetricLimit> &limit)
+    const std::optional<MDSPerfMetricLimit> &limit,
+    const bool reregister)
```
Comment (vshankar) [inline src/mgr/ActivePyModules.cc:1359]: nit: force_register?

File: src/mgr/MDSPerfMetricTypes.h [hunk 12]
```diff
@@ -313,6 +313,7 @@ std::ostream &operator<<(std::ostream &os, const MDSPerfMetricQuery &query);
 struct MDSPerfCollector : PerfCollector {
   std::map<MDSPerfMetricKey, PerformanceCounters> counters;
   std::set<mds_rank_t> delayed_ranks;
+  utime_t timestamp;
```
Comment (vshankar) [inline src/mgr/MDSPerfMetricTypes.h:316]: last_updated?

File: src/mgr/MetricCollector.cc [hunk 13]
```diff
@@ -22,30 +22,37 @@ MetricCollector<Query, Limit, Key, Report>::MetricCollector(MetricListener &list
 template <typename Query, typename Limit, typename Key, typename Report>
 MetricQueryID MetricCollector<Query, Limit, Key, Report>::add_query(
     const Query &query,
-    const std::optional<Limit> &limit) {
-  dout(20) << "query=" << query << ", limit=" << limit << dendl;
+    const std::optional<Limit> &limit,
+    const bool reregister) {
```
Comment (vshankar) [inline src/mgr/MetricCollector.cc:26]: fix alignment.

File: src/mgr/MetricCollector.cc [hunk 14]
```diff
@@ -22,30 +22,37 @@ MetricCollector<Query, Limit, Key, Report>::MetricCollector(MetricListener &list
...
-
-    query_id = next_query_id++;
-    auto it = queries.find(query);
-    if (it == queries.end()) {
-      it = queries.emplace(query, std::map<MetricQueryID, OptionalLimit>{}).first;
-      notify = true;
-    } else if (is_limited(it->second)) {
-      notify = true;
+    if (reregister) {
+      dout(20) << "query=" << query << ", limit=" << (limit ? stringify(*limit) : "unlimited") << dendl;
```
Comment (vshankar) [inline src/mgr/MetricCollector.cc:33]: duplicate logging in if..else block.

File: src/mgr/MetricCollector.cc [hunk 15]
```diff
@@ -99,7 +104,7 @@ int MetricCollector<Query, Limit, Key, Report>::remove_query(MetricQueryID query
 
 template <typename Query, typename Limit, typename Key, typename Report>
 void MetricCollector<Query, Limit, Key, Report>::remove_all_queries() {
-  dout(20) << dendl;
+  dout(20) << "removing all queries" << dendl;
```
Comment (vshankar) [inline src/mgr/MetricCollector.cc:107]: any reason to change the log message?

File: src/mgr/MetricCollector.cc [hunk 16]
```diff
@@ -22,30 +22,37 @@ MetricCollector<Query, Limit, Key, Report>::MetricCollector(MetricListener &list
...
-    query_id = next_query_id++;
-    auto it = queries.find(query);
-    if (it == queries.end()) {
-      it = queries.emplace(query, std::map<MetricQueryID, OptionalLimit>{}).first;
-      notify = true;
-    } else if (is_limited(it->second)) {
-      notify = true;
+    if (reregister) {
+      dout(20) << "query=" << query << ", limit=" << (limit ? stringify(*limit) : "unlimited") << dendl;
+      auto it = queries.find(query);
```
Comment (vshankar) [inline src/mgr/MetricCollector.cc:34]: How about refactoring this as::

```
auto it = queries.find(query);
if (it == queries.end()) {
    query_id = next_query_id++;
    it = queries.emplace(query, std::map<MetricQueryID, OptionalLimit>{}).first;
    notify = true;
} else if (reregister || is_limited(it->second)) {
    notify = true;
}

// note that emplace does not do anything if the key already exists
it->second.emplace(query_id, limit);
counters.emplace(query_id, std::map<Key, PerformanceCounters>{});
```

?

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 17]
```diff
@@ -164,6 +170,44 @@ def notify(self, cmdtag):
...
+            return
+        rank0_gid, state = gid_state
+        if (rank0_gid and (rank0_gid != self.prev_rank0_gid)):
+            if (state != 'up:active'):
+                return
+            #ronk0 MDS is up:active
+            ua_timestamp = datetime.now()
+            #Clear and wait for the empty metrics event before reregistering queries.
+            self.mx_event.clear()
+            self.mx_event.wait()
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:186]: Its not a  good idea to block during notify. Ideally, you should kick off the timer task when you see a new rank 0 mds.

In the timer task, check the last_updated timestamp with the one recorded when the new rank 0 mds was seen and then force register. If the last_updated timestamp is not refreshed, schedule the timer task again for the next second.

**Review by vshankar — CHANGES_REQUESTED** [review]

File: src/mgr/MetricCollector.cc (hunk 15 above)
Comment (joscollin) [reply src/mgr/MetricCollector.cc:107]: There is no logging there. So I put the log text.

File: src/mgr/ActivePyModules.cc (hunk 11 above)
Comment (joscollin) [reply src/mgr/ActivePyModules.cc:1359]: I think `reregister` should be appropriate word for the new parameters. We are not 'forcing' anything here, but just addressing something (handle the reregister event in `add_query()`) that was missing.

File: src/mgr/MetricCollector.cc (hunk 16 above)
Comment (joscollin) [reply src/mgr/MetricCollector.cc:34]: Then we need to find the query_id to be populated for the 'reregister' case. 

Your suggestion is good and I've considered this logic earlier, but then I dropped it not because of emplace but because the existing logic creates a new query_id (a new entry) everytime. But in the new reregister handler code, we just need to find something that's already existing in the map. Also I don't want to mess up with the 'is_limited' case, as the new query_id is being used for the 'is_limited' case too and then emplace it. So for the reregister handling, just finding the existing query_id and then do the handle_query_updated(). Hope this clarifies.

Comment (vshankar) [reply src/mgr/MetricCollector.cc:34]: OK. If you are concerned about `is_limited` part, you could add another `else {...}` block specific for re-register.

File: src/mgr/MetricCollector.cc (hunk 15 above)
Comment (vshankar) [reply src/mgr/MetricCollector.cc:107]: `dout(20) << dendl;` does. See `dout_prefix` in MetricCollect.cc source.

File: src/mgr/ActivePyModules.cc (hunk 11 above)
Comment (vshankar) [reply src/mgr/ActivePyModules.cc:1359]: `reregister` makes sense from mgr/stats context, however, in ceph-mgr context, it's "forcing" the register to be done unconditionally.

Comment (joscollin) [reply src/mgr/ActivePyModules.cc:1359]: Fixed the occurrences which are not specific to the `add_mds_perf_query` context. I hope that's better.

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 18]
```diff
@@ -166,6 +172,47 @@ def notify(self, cmdtag):
...
+        rank0_gid, state = gid_state
+        if (rank0_gid and (rank0_gid != self.prev_rank0_gid)):
+            if (state != 'up:active'):
+                return
+            #ronk0 MDS is up:active
+            ua_last_updated = datetime.now()
+            Timer(0, self.re_register_queries, args=(rank0_gid, ua_last_updated,)).start()
+
+    def re_register_queries(self, rank0_gid, ua_last_updated):
+        while True:
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:189]: This keeps the timer task pinned to this job - if there were other timer tasks queues, they would never get executed until this finishes. This should reschedule itself when it cannot proceed. Also, condition variables are to be used with a mutex.

**Review by vshankar — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 19]
```diff
@@ -166,6 +173,52 @@ def notify(self, cmdtag):
             self.log.debug("client_metadata={0}, to_purge={1}".format(
                 self.client_metadata['metadata'], self.client_metadata['to_purge']))
 
+    def notify_fsmap(self):
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:176]: shouldn't this be under `self.lock`?

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 20]
```diff
@@ -166,6 +173,52 @@ def notify(self, cmdtag):
...
+    def re_register_queries(self, rank0_gid, ua_last_updated):
+        #reregister queries if the metrics are the latest. Otherwise reschedule the timer and
+        #wait for the empty metrics
+        if self.mx_last_updated >= ua_last_updated:
+            self.log.debug("reregistering queries...")
+            for filter_spec in list(self.user_queries.keys()):
+                with self.lock:
+                    user_query = self.user_queries[filter_spec]
+                    self.log.debug("reregistering query: {}".format(user_query))
+                    user_query[QUERY_IDS] = self.register_mds_perf_query(filter_spec, True)
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:202]: nit: pass `reregister=True`

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 21]
```diff
@@ -166,6 +173,52 @@ def notify(self, cmdtag):
...
+                    user_query[GLOBAL_QUERY_ID] = self.register_global_perf_query(filter_spec, True)
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:203]: ditto.

File: src/pybind/mgr/stats/fs/perf_stats.py [hunk 22]
```diff
@@ -166,6 +173,52 @@ def notify(self, cmdtag):
...
+                                 self.re_register_queries, args=(rank0_gid, ua_last_updated,))
+            self.rqtimer.start()
+
+    def re_register_queries(self, rank0_gid, ua_last_updated):
+        #reregister queries if the metrics are the latest. Otherwise reschedule the timer and
+        #wait for the empty metrics
+        if self.mx_last_updated >= ua_last_updated:
+            self.log.debug("reregistering queries...")
+            for filter_spec in list(self.user_queries.keys()):
+                with self.lock:
```
Comment (vshankar) [inline src/pybind/mgr/stats/fs/perf_stats.py:199]: should this cover the entire check (i.e., before checking `self.mx_last_updated >= ua_last_updated`)?

**Review by vshankar — COMMENTED** [review]
minor nits. mostly looks good!

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test make check arm64

**Review by vshankar — COMMENTED** [review]
@joscollin - almost there.

This should be good for testing once the comments are addressed.

File: src/mgr/MDSPerfMetricTypes.h [hunk 23]
```diff
@@ -317,6 +317,7 @@ std::ostream &operator<<(std::ostream &os, const MDSPerfMetricQuery &query);
 struct MDSPerfCollector : PerfCollector {
   std::map<MDSPerfMetricKey, PerformanceCounters> counters;
   std::set<mds_rank_t> delayed_ranks;
+  utime_t last_updated;
```
[the remark this answers is not in the document]
Comment (joscollin) [reply src/mgr/MDSPerfMetricTypes.h:320]: We could use monotonic clock. But it would prevent the dumping, as mgr/PyFormatter doesn't support conversions from monotonic clock. No such issues with `utime_t`.

Please check: 
https://github.com/ceph/ceph/blob/8effd036db9c49d55875a5109c51b7497b99db35/src/mgr/ActivePyModules.cc#L1443
https://github.com/ceph/ceph/blob/master/src/mgr/PyFormatter.h#L89-L95
https://github.com/ceph/ceph/blob/8effd036db9c49d55875a5109c51b7497b99db35/src/pybind/mgr/stats/fs/perf_stats.py#L319

**Comment by joscollin** [comment]
jenkins test make check

File: src/mgr/MDSPerfMetricCollector.cc [hunk 24]
```diff
@@ -47,6 +47,7 @@ int MDSPerfMetricCollector::get_counters(PerfCollector *collector) {
 
   get_delayed_ranks(&c->delayed_ranks);
 
+  c->last_updated = ceph_clock_now();
```
Comment (vshankar) [inline src/mgr/MDSPerfMetricCollector.cc:50]: Why here? This need to be updated in `MDSPerfMetricCollector::process_reports()`

File: src/mgr/MetricCollector.cc [hunk 25]
```diff
@@ -114,6 +114,36 @@ void MetricCollector<Query, Limit, Key, Report>::remove_all_queries() {
   }
 }
 
+template <typename Query, typename Limit, typename Key, typename Report>
+int MetricCollector<Query, Limit, Key, Report>::reregister_query(MetricQueryID query_id) {
```
Comment (vshankar) [inline src/mgr/MetricCollector.cc:118]: You can do away with the query id since this is basically a brute-force register.

**Review by vshankar — CHANGES_REQUESTED** [review]

File: src/mgr/MDSPerfMetricTypes.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (joscollin) [reply src/mgr/MDSPerfMetricTypes.h:320]: I would like to use utime_t here, analyzing the complexity the suggested change would make on the mgr side and the time conversion operations in perf_stats.py. The other comments are fixed.

**Comment by vshankar** [comment]
@joscollin Please check commit https://github.com/ceph/ceph/commit/ff8007fd30ee21ca301ab6f86e4064fc5bdcedc8 - that switches to monotonic clocks.

File: src/mgr/MDSPerfMetricTypes.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (vshankar) [reply src/mgr/MDSPerfMetricTypes.h:320]: Monotonic clocks are preferred over realtime clocks for the very reason that it (monotonic) clocks can't go backwards. That's essential for things like such.

**Comment by joscollin** [comment]
@vshankar As we have avoided the query iterations in perf_stats now, A better name for the interface would be having something like `handle_query_` instead of `reregister_`. What do you say?

**Comment by vshankar** [comment]
> @vshankar As we have avoided the query iterations in perf_stats now, A better name for the interface would be having something like `handle_query_` instead of `reregister_`. What do you say?

Apologies for replying late.

handle_* generally indicate that the function is supposed to be invoked as a callback handler. I would stick to using reregister here.

**Comment by vshankar** [comment]
@joscollin please rebase.

**Comment by joscollin** [comment]
> > @vshankar As we have avoided the query iterations in perf_stats now, A better name for the interface would be having something like `handle_query_` instead of `reregister_`. What do you say?
> 
> Apologies for replying late.
> 
> handle_* generally indicate that the function is supposed to be invoked as a callback handler. I would stick to using reregister here.

alright.

**Comment by joscollin** [comment]
> @joscollin please rebase.

Done

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Mar-08
